src/signer/noscrypt_signer.cpp

```diff
@@ -0,0 +1,569 @@
+{
+    // TODO: Verify the incoming event.
+
+    // Extract and decrypt the event payload.
+    string encryptedContent = event->content;
+    string decryptedContent;
+
+    // NIP-04 encrypted strings include `?iv=` near the end (source: hodlbod).
+    if (encryptedContent.find("?iv=") != string::npos)
+    {
+        decryptedContent = this->_decryptNip04(encryptedContent);
+    }
+    else
+    {
+        decryptedContent = this->_decryptNip44(encryptedContent);
+    }
+
+    // Parse the decrypted string into a JSON object.
+    return decryptedContent;
+};
+
+inline shared_ptr<Filters> NoscryptSigner::_buildSignerMessageFilters() const
+{
+    auto filters = make_shared<Filters>();
+    filters->authors.push_back(this->_getRemotePublicKey());
+    filters->kinds.push_back(this->_nostrConnectKind);
+    filters->tags["p"] = { this->_getLocalPublicKey() };
+    filters->since = time(nullptr);
+
+    return filters;
+};
+
+promise<bool> NoscryptSigner::_pingSigner()
+{
+    promise<bool> pingPromise;
+
+    // Generate a ping message and wrap it for the signer.
+    nlohmann::json jrpc =
+    {
+        { "id", this->_generateSignerRequestId() },
+        { "method", "ping" },
+        { "params", nlohmann::json::array() }
+    };
+    auto messageEvent = this->_wrapSignerMessage(jrpc);
+
+    // Generate a filter to receive the response.
+    auto pingFilter = this->_buildSignerMessageFilters();
+
+    this->_nostrService->publishEvent(messageEvent);
+
+    // TODO: Handle the relay response.
+    this->_nostrService->queryRelays(
+        pingFilter,
+        [this, &pingPromise](const string&, shared_ptr<Event> pongEvent)
+        {
+            // 
+            string pongMessage = this->_unwrapSignerMessage(pongEvent);
+            pingPromise.set_value(pongMessage == "pong");
+        },
+        [&pingPromise](const string&)
+        {
+            pingPromise.set_value(false);
+        },
+        [&pingPromise](const string&, const string&)
+        {
+            pingPromise.set_value(false);
+        });
+    
+    return pingPromise;
+};
+
+#pragma endregion
+
+#pragma region Cryptography
+
+string NoscryptSigner::_encryptNip04(std::string input)
+{
+    throw runtime_error("NIP-04 encryption is not yet implemented.");
+};
+
+string NoscryptSigner::_decryptNip04(string input)
+{
+    throw runtime_error("NIP-04 decryption is not yet implemented.");
+};
+
+string NoscryptSigner::_encryptNip44(string input)
+{
+    NoscryptCipher cipher = NoscryptCipher(
+        NoscryptCipherVersion::NIP44,
+        NoscryptCipherMode::CIPHER_MODE_ENCRYPT
+    );
+
+	auto output = cipher.update(
+		this->_noscryptContext,
+		this->_localPrivateKey,
+		this->_remotePublicKey,
+		input
+    );
+
+    return output.empty()
+        ? string()
+        : NoscryptCipher::naiveEncodeBase64(output);
+};
+
+string NoscryptSigner::_decryptNip44(string input)
+{
+    //TODO handle input validation as per nip44 spec
```

Comment:
At some point I think well need to make sure to [check for a leading `#` character](https://github.com/paulmillr/nip44?tab=readme-ov-file#decryption) before base64 decoding. We can add that into the TODO section for validating payloads. 

I added this comment because it might be better to validate the message before calling this function, well have to decide what's better for error handling.

Reply:
Yeah let's tackle that in another PR.  Good suggestion, though.  Not a huge concern now, but we'll want to future-proof our code.